### ISSUE TITLE
Operator for getting cross of two channels with shared and duplicate keys

### ISSUE THREAD

**Issue by skptic** [nextflow-io/nextflow#253]
Trying to get the full cross of two channels based on a key (all items of ChannelA crossed with all items in ChannelB) is not so easy if there exists duplicate keys.

Given the following channels:

```
ch1 = Channel.from( ['datasetA',1], ['datasetA',2], ['datasetB',1], ['datasetB',2] )
ch2 = Channel.from( ['datasetA',1], ['datasetA',2], ['datasetB',1], ['datasetB',2] )

```

the required output would be:

```
datasetA, 1, 1
datasetA, 1, 2
datasetA, 2, 1
datasetA, 2, 2
datasetB, 1, 1
datasetB, 1, 2
datasetB, 2, 1
datasetB, 2, 2
```

Currently it is possible with the following:

```
ch2
  .groupTuple()
  .set {channel2_grouped}

channel2_grouped
  .cross(ch1)
  .flatMap{ alpha, omega ->  [ alpha[1], omega[1]].combinations().collect { [alpha[0], it[1], it[0]] }  }
  .println()
```

A function which performs the above would be ideal.

**Comment by pditommaso** [comment]
I'm giving a try to this. Basically you want a cartesian product for the tuples having a matching key. Is that right?

**Comment by pditommaso** [comment]
Resolved by #298.
